linkedlist/src/main/java/by/desckapg/linkedlist/LinkedList.java

```diff
@@ -0,0 +1,223 @@
+package by.desckapg.linkedlist;
+
+import java.util.NoSuchElementException;
+
+/**
+ * Generic doubly linked list implementation.
+ * Provides operations to add, retrieve, and remove elements by index
+ * and at the ends of the list. Not thread-safe.
+ * <p>
+ * Indexing:
+ * - Valid indices are in the range [0, size - 1].
+ * - add(index, value) inserts a new element before the element at the given index.
+ * To insert at the beginning or end, use addFirst(value) or addLast(value).
+ * <p>
+ * Complexity (amortized):
+ * - addFirst/addLast/removeFirst/removeLast: O(1)
+ * - get/add/remove by index: O(n)
+ *
+ * @param <T> the type of elements stored in this list
+ */
+public class LinkedList<T> {
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size = 0;
+
+    /**
+     * Returns the number of elements in this list.
+     *
+     * @return the list size
+     */
+    public int size() {
+        return size;
+    }
+
+    /**
+     * Returns the value of the first element.
+     *
+     * @return the first element's value
+     * @throws NoSuchElementException if the list is empty
+     */
+    public T getFirst() {
+        if (head == null) {
+            throw new NoSuchElementException();
+        }
+        return head.value;
+    }
+
+    /**
+     * Returns the value of the last element.
+     *
+     * @return the last element's value
+     * @throws NoSuchElementException if the list is empty
+     */
+    public T getLast() {
+        if (tail == null) {
+            throw new NoSuchElementException();
+        }
+        return tail.value;
+    }
+
+    /**
+     * Returns the value at the specified position in this list.
+     *
+     * @param index index of the element to return (0..size-1)
+     * @return the value at the specified position
+     * @throws IndexOutOfBoundsException if the index is out of range
+     */
+    public T get(int index) {
```

Comment:
Missing null check in get(). The loop in the get(int index) method doesn't properly handle the case where node becomes null during traversal.